ClipboardAPI/clipboard-change-event-explainer.md

```diff
@@ -0,0 +1,210 @@
+2. Not much useful unless the page focus requirement is removed from the async read/write clipboard API.
+3. Could lead to higher resource consumption due to continuous monitoring. E.g. if a large number of pages are listening to the clipboardchange event, when the clipboard is changed, then all the pages will receive the event at once, which might put load on the system.
+
+#### 5.2.3 Conclusion
+We favour Approach 1 - Page required to be in focus to receive event, since this approach reduces the possibility of misusing the clipboard change event in privacy related attacks.
+
+### 5.3 Event details 
+"clipboardchange" is a system level event, hence it is not bound to any DOM element. Therefore, there is no bubbling associated with the event. This event simply signifies that the system clipboard got changed. Hence performing “preventDefault” for this event has no effect since there are no actions associated with it. The event does not provide any additional arguments in the callback handler - it is expected that the web app calls Async API to read the clipboard and get the changed clipboard contents within the callback.
+
+## 6. Detailed design discussion
+
+### 6.1 Listen to clipboard change directly from the OS
+In this approach, the browser process can have a singleton service which listens to clipboard change event using OS specific APIs. If the platform doesn't provide any OS specific APIs, then we can consider polling to obtain the clipboard change event. 
+
+The browser process can then forward the clipboard change event to all listening renderer processes and fire the event after performing the page focus and user permissions check. In case the page is out of focus, the renderer process won't dispatch the event right away but will set an internal flag which will cause the renderer process to dispatch a "clipboardchange" event as soon as the page regains focus. 
+
+We can further optimize the page out of focus scenario as follows - when the page looses focus, we stop listening to clipboardchange event from the OS. Then upon regain of focus, we firstly check if the clipboard got changed by comparing clipboard hash using [SequenceNumber API](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getclipboardsequencenumber) and then fire a clipboardchange event if the hash is different from the previously stored one. Secondly, we restart listening to the clipboardchange event from the OS. This will help reduce system resource consumption since the browser won't need process every clipboard change event which is occurring in the OS, while the browser is not in focus.
+
+Moreover, we can ensure to only listen to the OS clipboard when there is at least one renderer process interested in the event.
+
+High level design doc available [here](https://docs.google.com/document/d/1bY2pzV6PSX56fiFcrXEgOjpFen07xaxmnsM5dqXFE1U/edit#heading=h.i1vomwqlf6kt)
```

Comment:
Explainers are the starting point for specifications. They shouldn't link to design docs or browser implementation details. This can be linked in the Chrome Status entry and I2P.